File: src/Tree/SymmetricTree.java
package Tree;

import MaximumDepthOfBinaryTree.TreeNode;


public class SymmetricTree {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(
                1,
                new TreeNode(2, new TreeNode(3), new TreeNode(4)),
                new TreeNode(2, new TreeNode(4), new TreeNode(3))
        );
        TreeNode root2 = new TreeNode(
                1,
                new TreeNode(2, null, new TreeNode(3)),
                new TreeNode(2, null, new TreeNode(3))
        );
        System.out.println(isSymmetric(root));
        System.out.println(isSymmetric(root2));
    }

    public static boolean isSymmetric(TreeNode root) {
        return dfs(root.left, root.right);
    }

    private static boolean dfs(TreeNode left, TreeNode right) {
        if (left == null && right == null) return true;

        if (left == null || right == null) return false;

        return left.val == right.val &&
                dfs(left.left, right.right) &&
                dfs(left.right, right.left);
    }
}

File: src/YandexList/SymmetricTree.kt
package YandexList

var ti = TreeNode(5)
var v = ti.value

class TreeNode(var value : Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}


class SymmetricTree {

    fun isSymmetric(root: TreeNode?): Boolean {
        if (root == null) {
            return true
        }
        return checkSymmetric(root.left, root.right)
    }

    private fun checkSymmetric(left: TreeNode?, right: TreeNode?): Boolean {
        if (left == null && right == null) {
            return true
        } else if (left != null && right != null) {
            return left.value == right.value && checkSymmetric(left.left, right.right) && checkSymmetric(left.right, right.left)
        }
        return false
    }
}